src/xclient_zmq_impl.cpp

```diff
@@ -0,0 +1,187 @@
+/***************************************************************************
+* Copyright (c) 2016, Johan Mabille, Sylvain Corlay, Martin Renou          *
+* Copyright (c) 2016, QuantStack                                           *
+*                                                                          *
+* Distributed under the terms of the BSD 3-Clause License.                 *
+*                                                                          *
+* The full license is in the file LICENSE, distributed with this software. *
+****************************************************************************/
+
+#include <iostream>
+
+#include "xeus-zmq/xauthentication.hpp"
+#include "xclient_zmq_impl.hpp"
+#include "xeus-zmq/xzmq_serializer.hpp"
+
+namespace xeus
+{
+
+    xclient_zmq_impl::xclient_zmq_impl(zmq::context_t& context,
+                                    const xeus::xconfiguration& config,
+                                    nl::json::error_handler_t eh)
+        : p_auth(make_xauthentication(config.m_signature_scheme, config.m_key))
+        , m_shell_client(context, config.m_transport, config.m_ip, config.m_shell_port)
+        , m_control_client(context, config.m_transport, config.m_ip, config.m_control_port)
+        , m_iopub_client(context, config)
+        , p_messenger(context)
+        , m_error_handler(eh)
+    {
+    }
+
+    // Has to be in the cpp because incomplete
+    // types are used in unique_ptr in the header
+    xclient_zmq_impl::~xclient_zmq_impl() = default;
+
+    void xclient_zmq_impl::send_on_shell(xmessage msg)
+    {
+        zmq::multipart_t wire_msg = xzmq_serializer::serialize(std::move(msg), *p_auth, m_error_handler);
+        m_shell_client.send_message(wire_msg);
+    }
+
+    void xclient_zmq_impl::send_on_control(xmessage msg)
+    {
+        zmq::multipart_t wire_msg = xzmq_serializer::serialize(std::move(msg), *p_auth, m_error_handler);
+        m_control_client.send_message(wire_msg);
+    }
+
+    std::optional<xmessage> xclient_zmq_impl::receive_on_shell(long timeout)
+    {
+        std::optional<zmq::multipart_t> wire_msg = m_shell_client.receive_message(timeout);
+
+        if (wire_msg.has_value())
+        {
+            return deserialize(wire_msg.value());
+        } else {
+            return std::nullopt;
+        }
+    }
+
+    std::optional<xmessage> xclient_zmq_impl::receive_on_control(long timeout)
+    {
+        std::optional<zmq::multipart_t> wire_msg = m_control_client.receive_message(timeout);
+
+        if (wire_msg.has_value())
+        {
+            return deserialize(wire_msg.value());
+        } else {
+            return std::nullopt;
+        }
+    }
+
+    void xclient_zmq_impl::register_shell_listener(const listener& l)
+    {
+        m_shell_listener = l;
+    }
+
+    void xclient_zmq_impl::register_control_listener(const listener& l)
+    {
+        m_control_listener = l;
+    }
+
+    std::size_t xclient_zmq_impl::iopub_queue_size() const
```

Comment:
Hasn't technically been put to use anywhere.
We can check for size and if size > 0 , then pop the message 

but `pop_iopub_message`, has been framed such that `if(!queue.empty)` then return message else `std::nullopt`, hence just by using `pop_iopub_message` we can get the message and proceed if we don't get back `std::nullopt`

Reply:
This method is meant to be used by the client. Sometimes you jst want to check that the queue size is not too big (because you have other messages to process for instance).